src/pk.c

```diff
@@ -6959,793 +7110,1199 @@ WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
+
+#if defined(WOLFSSL_QT) || defined(OPENSSL_ALL) \
+    || defined(WOLFSSL_OPENSSH) || defined(OPENSSL_EXTRA)
+
+/* Set the members of DhKey into WOLFSSL_DH
+ * Specify elements to set via the 2nd parameter
+ *
+ * @param [in, out] dh   DH key to synchronize.
+ * @param [in]      elm  Elements to synchronize.
+ * @return  1 on success.
+ * @return  -1 on failure.
+ */
+int SetDhExternal_ex(WOLFSSL_DH *dh, int elm)
 {
-    word32 idx = 0;
-    int    ret;
+    int ret = 1;
+    DhKey *key = NULL;
 
-    if (dh == NULL || dh->internal == NULL || derBuf == NULL || derSz <= 0) {
-        WOLFSSL_MSG("Bad function arguments");
-        return -1;
+    WOLFSSL_ENTER("SetDhExternal_ex");
+
+    /* Validate parameters. */
+    if ((dh == NULL) || (dh->internal == NULL)) {
+        WOLFSSL_ERROR_MSG("dh key NULL error");
+        ret = -1;
     }
 
-    ret = wc_DhKeyDecode(derBuf, &idx, (DhKey*)dh->internal, (word32)derSz);
-    if (ret < 0) {
-        WOLFSSL_MSG("wc_DhKeyDecode failed");
-        return -1;
+    if (ret == 1) {
+        /* Get the wolfSSL DH key. */
+        key = (DhKey*)dh->internal;
     }
-    dh->inSet = 1;
 
-    if (SetDhExternal(dh) != 1) {
-        WOLFSSL_MSG("SetDhExternal failed");
-        return -1;
+    if ((ret == 1) && (elm & ELEMENT_P)) {
+        /* Set the prime. */
+        if (SetIndividualExternal(&dh->p, &key->p) != 1) {
+            WOLFSSL_ERROR_MSG("dh param p error");
+            ret = -1;
+        }
+    }
+    if ((ret == 1) && (elm & ELEMENT_G)) {
+        /* Set the generator. */
+        if (SetIndividualExternal(&dh->g, &key->g) != 1) {
+            WOLFSSL_ERROR_MSG("dh param g error");
+            ret = -1;
+        }
+    }
+    if ((ret == 1) && (elm & ELEMENT_Q)) {
+        /* Set the order. */
+        if (SetIndividualExternal(&dh->q, &key->q) != 1) {
+            WOLFSSL_ERROR_MSG("dh param q error");
+            ret = -1;
+        }
+    }
+#ifdef WOLFSSL_DH_EXTRA
+    if ((ret == 1) && (elm & ELEMENT_PRV)) {
+        /* Set the private key. */
+        if (SetIndividualExternal(&dh->priv_key, &key->priv) != 1) {
+            WOLFSSL_ERROR_MSG("No DH Private Key");
+            ret = -1;
+        }
+    }
+    if ((ret == 1) && (elm & ELEMENT_PUB)) {
+        /* Set the public key. */
+        if (SetIndividualExternal(&dh->pub_key, &key->pub) != 1) {
+            WOLFSSL_ERROR_MSG("No DH Public Key");
+            ret = -1;
+        }
     }
+#endif /* WOLFSSL_DH_EXTRA */
 
-    return 1;
+    if (ret == 1) {
+        /* On success record that the external values have been set. */
+        dh->exSet = 1;
+    }
+
+    return ret;
+}
+/* Set the members of DhKey into WOLFSSL_DH
+ * DhKey was populated from wc_DhKeyDecode
+ * p, g, pub_key and priv_key are set.
+ *
+ * @param [in, out] dh   DH key to synchronize.
+ * @return  1 on success.
+ * @return  -1 on failure.
+ */
+int SetDhExternal(WOLFSSL_DH *dh)
+{
+    int elements = ELEMENT_P | ELEMENT_G | ELEMENT_PUB | ELEMENT_PRV;
```

Comment:
I assume not setting Q is intentional, but a comment here explaining why would be helpful.

Reply:
Comment added.